File: src/utils/videoEmbed.ts
/**
 * دعم الوسائط المرئية عبر التضمين (Embed) بدلاً من الرفع المباشر.
 *
 * السبب: رفع الملفات يحتاج Firebase Storage، وهو غير متاح في خطة Spark
 * المجانية. التضمين من يوتيوب أو Vimeo مجاني تماماً، ولا يستهلك تخزيناً
 * ولا نطاقاً من المنصة، ويحمّل أسرع للقارئ.
 */

export type VideoProvider = 'youtube' | 'vimeo' | 'unknown';

export interface ParsedVideo {
  provider: VideoProvider;
  videoId: string;
  embedUrl: string;
  thumbnailUrl: string;
}

/**
 * يستخرج معرّف الفيديو من رابط يوتيوب بكل صيغه المعروفة.
 */
function parseYouTube(url: string): string | null {
  const patterns = [
    /(?:youtube\.com\/watch\?v=)([A-Za-z0-9_-]{11})/,
    /(?:youtu\.be\/)([A-Za-z0-9_-]{11})/,
    /(?:youtube\.com\/embed\/)([A-Za-z0-9_-]{11})/,
    /(?:youtube\.com\/shorts\/)([A-Za-z0-9_-]{11})/
  ];
  for (const re of patterns) {
    const m = url.match(re);
    if (m) return m[1];
  }
  return null;
}

function parseVimeo(url: string): string | null {
  const m = url.match(/vimeo\.com\/(?:video\/)?(\d+)/);
  return m ? m[1] : null;
}

/**
 * يحلّل رابط فيديو ويعيد بيانات التضمين، أو null إن كان الرابط غير مدعوم.
 */
export function parseVideoUrl(url: string): ParsedVideo | null {
  if (!url || !url.trim()) return null;
  const clean = url.trim();

  const ytId = parseYouTube(clean);
  if (ytId) {
    return {
      provider: 'youtube',
      videoId: ytId,
      // rel=0 يقلل اقتراحات الفيديوهات الخارجية بعد الانتهاء
      embedUrl: `https://www.youtube-nocookie.com/embed/${ytId}?rel=0`,
      thumbnailUrl: `https://img.youtube.com/vi/${ytId}/hqdefault.jpg`
    };
  }

  const vimeoId = parseVimeo(clean);
  if (vimeoId) {
    return {
      provider: 'vimeo',
      videoId: vimeoId,
      embedUrl: `https://player.vimeo.com/video/${vimeoId}`,
      thumbnailUrl: ''
    };
  }

  return null;
}

export function isValidVideoUrl(url: string): boolean {
  return parseVideoUrl(url) !== null;
}

/**
 * يتحقق من أن رابط الصورة صالح الشكل.
 * لا نرفع الصور — تُستخدم روابط خارجية فقط.
 */
export function isValidImageUrl(url: string): boolean {
  if (!url || !url.trim()) return false;
  try {
    const u = new URL(url.trim());
    return u.protocol === 'https:' || u.protocol === 'http:';
  } catch {
    return false;
  }
}

File: android-native/app/src/main/kotlin/studio/ai/literium/literium_app/ui/components/VideoEmbed.kt
package studio.ai.literium.literium_app.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.unit.dp

data class ParsedVideo(val provider: String, val videoId: String, val embedUrl: String)

private val YOUTUBE_PATTERNS = listOf(
    Regex("""(?:youtube\.com/watch\?v=)([A-Za-z0-9_-]{11})"""),
    Regex("""(?:youtu\.be/)([A-Za-z0-9_-]{11})"""),
    Regex("""(?:youtube\.com/embed/)([A-Za-z0-9_-]{11})"""),
    Regex("""(?:youtube\.com/shorts/)([A-Za-z0-9_-]{11})""")
)
private val VIMEO_PATTERN = Regex("""vimeo\.com/(?:video/)?(\d+)""")

/** Kotlin port of `src/utils/videoEmbed.ts`'s `parseVideoUrl` — same YouTube/Vimeo pattern set, same
 *  `youtube-nocookie.com`/`player.vimeo.com` embed URLs. */
fun parseVideoUrl(url: String): ParsedVideo? {
    val clean = url.trim()
    if (clean.isEmpty()) return null

    for (pattern in YOUTUBE_PATTERNS) {
        pattern.find(clean)?.groupValues?.get(1)?.let { id ->
            return ParsedVideo("youtube", id, "https://www.youtube-nocookie.com/embed/$id?rel=0")
        }
    }
    VIMEO_PATTERN.find(clean)?.groupValues?.get(1)?.let { id ->
        return ParsedVideo("vimeo", id, "https://player.vimeo.com/video/$id")
    }
    return null
}

/** Production origin the video-embed HTML wrapper is loaded with — see [VideoEmbed]'s KDoc for why
 *  this, not `null`, is required for YouTube playback to work at all. */
private const val PRODUCTION_ORIGIN = "https://literium.ai.studio/"

/**
 * Kotlin port of `VideoEmbed.tsx` — a YouTube/Vimeo `<iframe>` embed, reserved at a 16:9 aspect ratio
 * so surrounding layout never jumps once it loads. Renders nothing for an unparseable [url], matching
 * source's own `if (!parsed) return null`.
 *
 * Wraps [ParsedVideo.embedUrl] in a real `<iframe>` inside a tiny host page, loaded via
 * [IsolatedWebView]'s `html`+`baseUrl` path rather than navigating the WebView directly to the embed
 * URL (`url =`) — a real-device bug report ("YouTube error 153: خطأ في إعدادات مشغل الفيديو"): loading
 * the embed URL as the WebView's own top-level document, the way the `url` path does, gives YouTube's
 * player no parent-frame/origin context at all (effectively `about:blank`), which YouTube's embed
 * player rejects outright. `VideoEmbed.tsx` on web never hits this because its `<iframe src=embedUrl>`
 * is genuinely nested inside the real literium.ai.studio page — this reproduces that exact structure
 * natively, with [PRODUCTION_ORIGIN] as the iframe's real parent origin.
 */
@Composable
fun VideoEmbed(url: String, modifier: Modifier = Modifier) {
    val parsed = parseVideoUrl(url) ?: return
    val html = """
        <!DOCTYPE html><html><head><meta charset="utf-8">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        <style>html,body{margin:0;padding:0;background:#0f172a;overflow:hidden}
        iframe{position:absolute;inset:0;width:100%;height:100%;border:0}</style>
        </head><body>
        <iframe src="${parsed.embedUrl}" allow="accelerometer; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
        </body></html>
    """.trimIndent()
    IsolatedWebView(
        html = html,
        baseUrl = PRODUCTION_ORIGIN,
        modifier = modifier
            .fillMaxWidth()
            .aspectRatio(16f / 9f)
            .clip(RoundedCornerShape(12.dp))
            .background(MaterialTheme.colorScheme.surfaceVariant)
    )
}
